# **Compiled Vs Interpretted**



**Compiled** :
* Preprocess the instructions and create an executable which is ready to execute.

* Faster. Important when speed of execution is critical

* Compilation step is platform dependent


**Interpretted**:

* Processes instructions line by line during execution. Don't have build step.

* NO PLATORM specific preprocessing.


-----------------------------------------


Python is using Hybrid Approach.

**Step 1** : High level instructions(*.py) -> Byte-code equivalent(*.pyc)
* Platform independent

**Step -2** : Byte-code Instruction -> Python Virtual Machine(PVM)->Interpret and execute


-------------------------------------------

Code blocks segragated using White space.



# **Storage and Access**



* Python data is represented by **Objects**
* Every python Object has 3 things associated with it :

      1.  Value
      2.  Type
      3.  ID


* Python doen't store pointers. So we cannot use any memory address to access some value.

* Python object is stored in **Heap**

* Based on where object is stored on heap, its ID value is determined and its associated with that oject throughout life time.




In [1]:
x = 100
y = 100.5
z = 100
print(x,y,z)
print(id(x),id(y),id(z))
print(type(x),type(y),type(z))


100 100.5 100
10760904 133261791118768 10760904
<class 'int'> <class 'float'> <class 'int'>


* Above code shows x and z objects are associated with same ID.

* When we say x=100, 100 is an object and x is label pointing to 100

   So when we say y=100, there is another label created that points to 100.

* IF we perfrom, x=1,y=1,z=1, then noone is pointing to 100 so object 100 will be deleted from heap.

In [2]:
x = 100
y = x
print(x,y)
x = 5
print(x,y)

a = [1, 2, 3, 4]
b = a
print(a,b)

a[0] = 100
print(a,b)


100 100
5 100
[1, 2, 3, 4] [1, 2, 3, 4]
[100, 2, 3, 4] [100, 2, 3, 4]


In above example, as list is a mutable object, if we change after its creation, change happens at same place. But values(x,y) are not mutable. So when we change x, it startes refereing to another object.


## List of common **built-in object types** in Python

int  
float  
complex  
bool  
str  
list  
tuple  
set  
frozenset  
dict  
bytes  
bytearray  
memoryview  
range  
NoneType  
EllipsisType  
NotImplementedType  
function  
lambda  
generator  
coroutine  
async_generator  
module  
class  
classmethod  
staticmethod  
property  
object  
type  
slice  
map  
filter  
zip  
enumerate  
reversed  
super  
Exception  
BaseException  
file (via open)  

# **Mutability**

Object is said to be mutable, if it can be changed post its creation.


**List of Mutable Objects:**

list  
dict  
set  
bytearray


In [3]:
print(a[1])
print(id(a))
print(id(a[1]))
a[1] = 200

print(a[1])
print(id(a))
print(id(a[1]))

2
133261796176256
10757768
200
133261796176256
10764104


* As per above example, ID of list remains same, even after change of values in list.

* ID of list elements changes with value change


In [6]:
#Number of Labels present for Value 1
import sys
print(sys.getrefcount(1))

1000006290


* There are 1000006290 references to Object "1"

## List of Immutable Objects:
int  
float  
complex  
bool  
str  
tuple  
frozenset  
bytes  
range  
NoneType  


# **Identifiers**



* Same as C - Start with alphabet,digits allowed afterwards


In [8]:
help("keywords")
help("and")
#1.11.35


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 

Boolean operations
******************

   or_test  ::= and_test | or_test "or" and_test
   and_test ::= not_test | and_test "and" not_test
   not_test ::= comparison | "not" not_test

In the context of Boolean operations, and also when expressions are
used by control flow statements, the following values are interpreted
as false: "False

# **Modes of operation : Interactive and Script**

In [9]:
x = "python"
dir(x)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

In [10]:
print(x)
x.upper()


python


'PYTHON'

In [11]:

x = []
bool(x)

False

In [14]:
#Multi-line string
x = """ .py file (your code) → compiled into bytecode (.pyc)
*Platform independent
*This happens automatically during import or first run

Bytecode is sent to PVM (Python Virtual Machine)
*PVM interprets and runs line-by-line
*PVM itself is platform-dependent
"""
print(x)

 .py file (your code) → compiled into bytecode (.pyc)
*Platform independent
*This happens automatically during import or first run

Bytecode is sent to PVM (Python Virtual Machine)
*PVM interprets and runs line-by-line
*PVM itself is platform-dependent



# **String Operations**


String is :
 Immutable
 Iterable


In [16]:
s = "Wipro Technologies"
print(s[3])
print(s[-1])
print(s[-3])
print(s[0])
print(s[1])
print(s[-1])
print(len(s))

r
s
i
W
i
s
18


In [17]:
#slicing
#[start,end,step]
print(s[::1])
print(s[6:9:1]) #9 is upto and non-inclusive
print(s[::-1]) # Most impotant : Reerse a string


Wipro Technologies
Tec
seigolonhceT orpiW


---

### 🔤 **Basic String Operations**

| Operation           | Description        | Example                                |
| ------------------- | ------------------ | -------------------------------------- |
| `+` (concatenation) | Joins two strings  | `"Hello" + " World"` → `'Hello World'` |
| `*` (repetition)    | Repeats string     | `"Hi" * 3` → `'HiHiHi'`                |
| `in`                | Checks substring   | `'lo' in 'Hello'` → `True`             |
| `len()`             | Gets length        | `len("Python")` → `6`                  |
| `str()`             | Converts to string | `str(123)` → `'123'`                   |

---

### ✂️ **String Slicing & Indexing**

| Operation  | Description                 | Example                   |
| ---------- | --------------------------- | ------------------------- |
| `s[i]`     | Character at index `i`      | `"abc"[1]` → `'b'`        |
| `s[i:j]`   | Substring from `i` to `j-1` | `"python"[1:4]` → `'yth'` |
| `s[i:j:k]` | Slicing with step `k`       | `"abcdef"[::2]` → `'ace'` |

---

### 🧼 **String Methods**

| Method               | Description                       | Example                                   |
| -------------------- | --------------------------------- | ----------------------------------------- |
| `.lower()`           | Lowercase                         | `'Hello'.lower()` → `'hello'`             |
| `.upper()`           | Uppercase                         | `'hello'.upper()` → `'HELLO'`             |
| `.capitalize()`      | First letter uppercase            | `'python'.capitalize()` → `'Python'`      |
| `.title()`           | Each word capitalized             | `'hello world'.title()` → `'Hello World'` |
| `.strip()`           | Remove surrounding spaces         | `' hello '.strip()` → `'hello'`           |
| `.replace(old, new)` | Replace substring                 | `'apple'.replace('p', 'b')` → `'abble'`   |
| `.split(sep)`        | Split into list                   | `'a,b,c'.split(',')` → `['a', 'b', 'c']`  |
| `.join(iterable)`    | Join elements with separator      | `'--'.join(['a', 'b'])` → `'a--b'`        |
| `.find(sub)`         | Index of first occurrence         | `'banana'.find('na')` → `2`               |
| `.count(sub)`        | Count occurrences                 | `'banana'.count('a')` → `3`               |
| `.startswith(sub)`   | Checks start                      | `'Python'.startswith('Py')` → `True`      |
| `.endswith(sub)`     | Checks end                        | `'script.py'.endswith('.py')` → `True`    |
| `.isdigit()`         | Checks if all chars are digits    | `'123'.isdigit()` → `True`                |
| `.isalpha()`         | Checks if all chars are alphabets | `'abc'.isalpha()` → `True`                |

---



---

### 🔤 **Python String Operations – Cheat Sheet**

| Operation / Method   | Description                             | Example                                   |
| -------------------- | --------------------------------------- | ----------------------------------------- |
| `+` (concatenation)  | Joins two strings                       | `"Hello" + " World"` → `"Hello World"`    |
| `*` (repetition)     | Repeats the string                      | `"Hi" * 3` → `"HiHiHi"`                   |
| `in` (membership)    | Checks if substring exists              | `'lo' in 'Hello'` → `True`                |
| `len()`              | Gets length of string                   | `len('Python')` → `6`                     |
| `str()`              | Converts to string                      | `str(123)` → `'123'`                      |
| `s[i]`               | Character at index `i`                  | `'abc'[1]` → `'b'`                        |
| `s[i:j]`             | Substring from i to j-1                 | `'python'[1:4]` → `'yth'`                 |
| `s[i:j:k]`           | Substring with step `k`                 | `'abcdef'[::2]` → `'ace'`                 |
| `.lower()`           | Converts to lowercase                   | `'Hello'.lower()` → `'hello'`             |
| `.upper()`           | Converts to uppercase                   | `'hello'.upper()` → `'HELLO'`             |
| `.capitalize()`      | Capitalizes first character             | `'python'.capitalize()` → `'Python'`      |
| `.title()`           | Capitalizes each word                   | `'hello world'.title()` → `'Hello World'` |
| `.strip()`           | Trims surrounding whitespace            | `' hello '.strip()` → `'hello'`           |
| `.replace(old, new)` | Replaces substring                      | `'apple'.replace('p', 'b')` → `'abble'`   |
| `.split(sep)`        | Splits string by separator              | `'a,b,c'.split(',')` → `['a', 'b', 'c']`  |
| `.join(iterable)`    | Joins list into string                  | `'--'.join(['a', 'b'])` → `'a--b'`        |
| `.find(sub)`         | Finds index of substring                | `'banana'.find('na')` → `2`               |
| `.count(sub)`        | Counts substring occurrences            | `'banana'.count('a')` → `3`               |
| `.startswith(sub)`   | Checks if string starts with substring  | `'Python'.startswith('Py')` → `True`      |
| `.endswith(sub)`     | Checks if string ends with substring    | `'script.py'.endswith('.py')` → `True`    |
| `.isdigit()`         | Checks if all characters are digits     | `'123'.isdigit()` → `True`                |
| `.isalpha()`         | Checks if all characters are alphabetic | `'abc'.isalpha()` → `True`                |

---




In [18]:
str = "python is a simple language."
str.capitalize()

'Python is a simple language.'

In [19]:
str.title()

'Python Is A Simple Language.'

In [20]:
str.upper()

'PYTHON IS A SIMPLE LANGUAGE.'

In [25]:
str2 = "Python Is a Simple Language."
print(str,str2)

if(str==str2):
  print("Both are same")
else:
  print("Both are different")


python is a simple language. Python Is a Simple Language.
Both are different


In [28]:
str2 = "Python Is a Simple Language."
print(str,str2)
str3 = str.casefold()
str4 = str2.casefold()
print(str3,str4)

if(str3==str4):
  print("Both are same")
else:
  print("Both are different")

python is a simple language. Python Is a Simple Language.
python is a simple language. python is a simple language.
Both are same


### 🔤 `.lower()`

* Converts only standard uppercase ASCII characters (A–Z) to lowercase.
* **Best for English** and basic lowercase conversion.

#### 📦 Example:

```python
'Hello'.lower()     → 'hello'
'PYTHON'.lower()    → 'python'
```

---

### 🌐 `.casefold()`

* More **aggressive and Unicode-aware** version of `.lower()`
* Used for **caseless matching**, especially with **non-English characters**
* Recommended when **comparing strings** (e.g. in search, comparison)

#### 📦 Example with Unicode:

```python
'straße'.lower()     → 'straße'  
'straße'.casefold()  → 'strasse'  ✅
```

("ß" in German is equivalent to "ss" in lowercase comparisons)

---

### ✅ Summary Table

| Method        | Use Case                        | Unicode-aware | Aggressiveness |
| ------------- | ------------------------------- | ------------- | -------------- |
| `.lower()`    | Basic lowercase (e.g., English) | ❌ No          | 🔹 Normal      |
| `.casefold()` | For comparisons, all languages  | ✅ Yes         | 🔸 Strong      |

---
